File: src/backend/app/schemas/knowledge_graph.py
"""
知识图谱相关的 Pydantic Schemas
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ==================== 知识图谱创建 ====================

class KnowledgeGraphCreate(BaseModel):
    """创建知识图谱请求"""
    
    name: str = Field(..., min_length=1, max_length=100, description="图谱名称")
    description: Optional[str] = Field(None, description="图谱描述")
    subject: Optional[str] = Field(None, max_length=50, description="学科分类")
    cover_image_url: Optional[str] = Field(None, max_length=500, description="封面图片URL")
    is_public: Optional[bool] = Field(False, description="是否公开")


# ==================== 知识图谱更新 ====================

class KnowledgeGraphUpdate(BaseModel):
    """更新知识图谱请求"""
    
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="图谱名称")
    description: Optional[str] = Field(None, description="图谱描述")
    subject: Optional[str] = Field(None, max_length=50, description="学科分类")
    cover_image_url: Optional[str] = Field(None, max_length=500, description="封面图片URL")
    is_public: Optional[bool] = Field(None, description="是否公开")


# ==================== 知识图谱响应 ====================

class KnowledgeGraphBase(BaseModel):
    """知识图谱基础信息"""
    
    id: UUID
    name: str
    description: Optional[str] = None
    subject: Optional[str] = None
    cover_image_url: Optional[str] = None
    is_public: bool
    node_count: int
    edge_count: int
    created_at: datetime
    updated_at: datetime
    
    model_config = {"from_attributes": True}


class KnowledgeGraphResponse(KnowledgeGraphBase):
    """知识图谱响应"""
    
    pass


class KnowledgeGraphDetailResponse(KnowledgeGraphBase):
    """知识图谱详细信息响应"""
    
    user_id: UUID
    total_review_count: int
    last_accessed_at: Optional[datetime] = None
    
    model_config = {"from_attributes": True}


# ==================== 知识图谱列表 ====================

class KnowledgeGraphListItem(BaseModel):
    """知识图谱列表项"""
    
    id: UUID
    name: str
    description: Optional[str] = None
    subject: Optional[str] = None
    cover_image_url: Optional[str] = None
    node_count: int
    is_public: bool
    created_at: datetime
    updated_at: datetime
    
    model_config = {"from_attributes": True}


# ==================== 知识图谱统计 ====================

class KnowledgeGraphStats(BaseModel):
    """知识图谱统计信息"""
    
    total_nodes: int = Field(..., description="总节点数")
    total_relations: int = Field(..., description="总关联数")
    total_tags: int = Field(..., description="总标签数")
    review_due_count: int = Field(..., description="待复习节点数")
    last_review_at: Optional[datetime] = Field(None, description="最后复习时间")

File: src/frontend/src/services/knowledgeGraph.ts
import apiClient from './api';

export type UUID = string;

// 类型定义
export interface KnowledgeGraph {
  id: UUID;
  user_id: UUID;
  name: string;
  description?: string;
  subject?: string;
  cover_image_url?: string;
  is_public: boolean;
  node_count: number;
  created_at: string;
  updated_at: string;
}

export interface CreateGraphRequest {
  name: string;
  description?: string;
  subject?: string;
  cover_image_url?: string;
  is_public?: boolean;
}

export interface UpdateGraphRequest {
  name?: string;
  description?: string;
  subject?: string;
  cover_image_url?: string;
  is_public?: boolean;
}

export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface GraphStatistics {
  graph_id: UUID;
  total_nodes: number;
  total_relations: number;
  total_tags: number;
  review_due_count: number;
  last_review_at?: string;
}

// 知识图谱服务
export const knowledgeGraphService = {
  // 创建知识图谱
  async createGraph(data: CreateGraphRequest): Promise<KnowledgeGraph> {
    const response = await apiClient.post<KnowledgeGraph>('/api/v1/graphs/', data);
    return response.data;
  },

  // 获取知识图谱列表
  async getGraphs(page: number = 1, pageSize: number = 10): Promise<PaginatedResponse<KnowledgeGraph>> {
    const response = await apiClient.get<PaginatedResponse<KnowledgeGraph>>('/api/v1/graphs/', {
      params: { page, page_size: pageSize },
    });
    return response.data;
  },

  // 获取知识图谱详情
  async getGraph(graphId: UUID): Promise<KnowledgeGraph> {
    const response = await apiClient.get<KnowledgeGraph>(`/api/v1/graphs/${graphId}`);
    return response.data;
  },

  // 更新知识图谱
  async updateGraph(graphId: UUID, data: UpdateGraphRequest): Promise<KnowledgeGraph> {
    const response = await apiClient.put<KnowledgeGraph>(`/api/v1/graphs/${graphId}`, data);
    return response.data;
  },

  // 删除知识图谱
  async deleteGraph(graphId: UUID): Promise<void> {
    await apiClient.delete(`/api/v1/graphs/${graphId}`);
  },

  // 获取图谱统计信息
  async getGraphStatistics(graphId: UUID): Promise<GraphStatistics> {
    const response = await apiClient.get<GraphStatistics>(`/api/v1/graphs/${graphId}/stats`);
    return response.data;
  },
};

